cpp/src/compaction/stream_compaction_ops.cu

```diff
@@ -224,109 +230,232 @@ gdf_error gpu_apply_stencil(gdf_column *lhs, gdf_column * stencil, gdf_column *
-	);
+		typedef thrust::transform_iterator<is_bit_set, zipped_mask > bit_set_iterator;
+		bit_set_iterator bit_set_iter = thrust::make_transform_iterator<is_bit_set,zipped_mask>(
+				zipped_mask_iter,
+				is_bit_set()
+		);
 
-	//copy the bitmask to device_vector of int8
-	thrust::copy(exec, bit_set_iter, bit_set_iter + num_values, valid_bit_mask.begin());
+		//copy the bitmask to device_vector of int8
+		thrust::copy(exec, bit_set_iter, bit_set_iter + num_values, valid_bit_mask.begin());
 
-	//remove the values that don't pass the stencil
-	thrust::remove_if(exec,valid_bit_mask.begin(), valid_bit_mask.begin() + num_values,zipped_stencil_iter, is_stencil_true<thrust::detail::normal_iterator<thrust::device_ptr<int8_t> >::value_type >());
+		//remove the values that don't pass the stencil
+		thrust::remove_if(exec,valid_bit_mask.begin(), valid_bit_mask.begin() + num_values,zipped_stencil_iter_one, is_stencil_true_one<thrust::detail::normal_iterator<thrust::device_ptr<int8_t> >::value_type >());
 
-	//recompact the values and store them in the output bitmask
-	//we can group them into pieces of 8 because we aligned this earlier on when we made the device_vector
-	thrust::detail::normal_iterator<thrust::device_ptr<int64_t> > valid_bit_mask_group_8_iter =
-			thrust::detail::make_normal_iterator(thrust::device_pointer_cast((int64_t *) valid_bit_mask.data().get()));
+		//recompact the values and store them in the output bitmask
+		//we can group them into pieces of 8 because we aligned this earlier on when we made the device_vector
+		thrust::detail::normal_iterator<thrust::device_ptr<int64_t> > valid_bit_mask_group_8_iter =
+				thrust::detail::make_normal_iterator(thrust::device_pointer_cast((int64_t *) valid_bit_mask.data().get()));
 
 
-	//you may notice that we can write out more bytes than our valid_num_bytes, this only happens when we are not aligned to  GDF_VALID_BITSIZE bytes, becasue the
-	//arrow standard requires 64 byte alignment, this is a safe assumption to make
-	thrust::transform(exec, valid_bit_mask_group_8_iter, valid_bit_mask_group_8_iter + ((num_values + GDF_VALID_BITSIZE - 1) / GDF_VALID_BITSIZE),
-			thrust::detail::make_normal_iterator(thrust::device_pointer_cast(output->valid)),bit_mask_pack_op());
+		//you may notice that we can write out more bytes than our valid_num_bytes, this only happens when we are not aligned to  GDF_VALID_BITSIZE bytes, becasue the
+		//arrow standard requires 64 byte alignment, this is a safe assumption to make
+		thrust::transform(exec, valid_bit_mask_group_8_iter, valid_bit_mask_group_8_iter + ((num_values + GDF_VALID_BITSIZE - 1) / GDF_VALID_BITSIZE),
+				thrust::detail::make_normal_iterator(thrust::device_pointer_cast(output->valid)),bit_mask_pack_op());
+	}
+	else {
+		//temp storage for cub call
+		int  *d_num_selected_out;
+		cudaMalloc(&d_num_selected_out,sizeof(int));
+		void     *d_temp_storage = NULL;
+		size_t   temp_storage_bytes = 0;
+		int out_size;
+		if(width == 1){
```

Comment:
This should use the type_dispatcher instead of this manual if/else on the bit-width.

https://github.com/rapidsai/cudf/blob/master/cpp/src/dataframe/type_dispatcher.hpp#L89

Reply:
Removed after refactoring.